### PR TITLE
Migrate synthetic data generation from iris-ugrid

### PR DESCRIPTION
**Pull request by trexfeathers**
### Description

For original review, see SciTools-incubator/iris-ugrid#38.

I'm keen in future to extend this beyond UGRID-NetCDF files, as the pattern is very useful for good benchmark coverage.

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

File: MANIFEST.in [hunk 1]
```diff
@@ -6,6 +6,7 @@ recursive-include lib/iris/tests/results *.cml *.cdl *.txt *.xml *.json
 recursive-include lib/iris/etc *
 include lib/iris/fileformats/_pyke_rules/*.k?b
 include lib/iris/tests/stock*.npz
```
Comment (pp-mo) [inline MANIFEST.in:8]: Quite unrelated really, but I see that this line is now obsolete + should probably have been removed in #3913
```suggestion
```

File: lib/iris/tests/unit/tests/stock/test_netcdf.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+    @classmethod
+    def tearDownClass(cls):
+        # Destroy the temp directory.
+        shutil.rmtree(cls.temp_dir)
+
+    def create_synthetic_file(self, **create_kwargs):
+        # Should be overridden to invoke one of the create_file_ functions.
+        # E.g.
+        # return create_file__xios_2d_face_half_levels(
```
Comment (pp-mo) [inline lib/iris/tests/unit/tests/stock/test_netcdf.py:34]: Just for consistency..
```suggestion
        # return netcdf.create_file__xios_2d_face_half_levels(
```

File: lib/iris/tests/unit/tests/stock/test_netcdf.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+    def check_cube(self, cube, shape, location, level):
+        # Basic checks on the primary data cube.
+        self.assertEqual(cube.var_name, "thing")
+        self.assertEqual(cube.long_name, "thingness")
+        self.assertEqual(cube.shape, shape)
+
+        # Also a few checks on the attached mesh-related information.
+        last_dim = cube.ndim - 1
+        self.assertTrue(hasattr(cube, "mesh"))
```
Comment (pp-mo) [inline lib/iris/tests/unit/tests/stock/test_netcdf.py:53]: This bit isn't quite right.  A cube now _always_ has a "mesh" attribute.
You should probably check either that it is non-None, or maybe that it is a `Mesh` ?

**Review by pp-mo — CHANGES_REQUESTED** [review]
Comparing to the content in ris-ugrid, I see you subtly refactored the scope of this with a view to future generalisation, e.g. beyond netcdf.  Which I think is all good, and nicely done!

I thought of some tiny tweaks, but otherwise looks fine.

File: lib/iris/tests/unit/tests/stock/test_netcdf.py (hunk 3 above)
Comment (trexfeathers) [reply lib/iris/tests/unit/tests/stock/test_netcdf.py:53]: Ah! This has already caught me out somewhere else, and now I know why! Will fix...

Comment (trexfeathers) [reply lib/iris/tests/unit/tests/stock/test_netcdf.py]: 736a6216e095cb483314457d4ab8bedf3ef2fa7e

File: lib/iris/tests/unit/tests/stock/test_netcdf.py (hunk 2 above)
Comment (trexfeathers) [reply lib/iris/tests/unit/tests/stock/test_netcdf.py]: 72538e7bd8586676944df8272693937b374be9a9

**Comment by trexfeathers** [comment]
Thanks @pp-mo !

**Pull request merged**